### PR TITLE
IGNITE-15208 Remove unnecessary rebalance order classes for exchange thread

### PR DESCRIPTION
**Pull request by Mmuzaf**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 1]
```diff
@@ -4028,61 +3960,32 @@ void trim() {
...
-         * Creates a new instance of CacheRebalanceOrder.
-         *
-         * @param order Cache rebalance order.
-         * @param mode Cache rebalance mode.
-         */
-        public CacheRebalanceOrder(int order, CacheRebalanceMode mode) {
-            this.order = order;
-            this.mode = mode;
-        }
+    private static class CacheRebalanceOrderComparator implements Comparator<CacheGroupContext>, Serializable {
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java:3963]: Yes, this is a common practice to implement `Serializable` interface by a comparator.
In this particular case, `assignsSet` is always a "local map" and will never be serialized/deserialized.
So, it is not required to implement `Serializable`

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheGroupContext.java [hunk 2]
```diff
@@ -1315,4 +1316,22 @@ public void removeIOStatistic(boolean destroy) {
         if (statHolderIdx != IoStatisticsHolderNoOp.INSTANCE)
             ctx.kernalContext().metric().remove(statHolderIdx.metricRegistryName(), destroy);
     }
+
+    /** {@inheritDoc} */
+    @Override public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheGroupContext.java:1321]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java (hunk 1 above)
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java [hunk 3]
```diff
@@ -305,9 +306,6 @@
     /** Histogram of blocking PME durations. */
     private volatile HistogramMetricImpl blockingDurationHistogram;
 
-    /** Delay before rebalancing code is start executing after exchange completion. For tests only. */
```
[the remark this answers is not in the document]
Comment (Mmuzaf) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java:308]: This parameter is used only for tests (to delay the exchange thread) and doesn't relate to the rebalance procedure itself. I agree with you that the naming of this property is confusing, so it's better to remove it at all.

The 'right' rebalance delay still available through public API.
org.apache.ignite.client.ClientCacheConfiguration#getRebalanceDelay

[the remark this answers is not in the document]
Comment (sk0x50) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePartitionExchangeManager.java:308]: Yes, that is my fault. You are absolutely right. Thank you for the clarification.

**Comment by Mmuzaf** [comment]
@sk0x50 If everything is OK, can I proceed with the merge?

**Pull request merged**